docs: Commands code sharing design document (#1135)
platform-platform/monorepo: metrics/ci_integrations/docs/05_sync_and_validate_commands_code_sharing.md

## Proposed Change

1. Create a `FileReader` and a `FileHelper` classes that provide methods for working with the file system.
2. Create a `RawIntegrationConfigFactory` class that creates the `RawIntegrationConfig` by the provided config file path, and inject it into a `ValidateCommand`.
3. Add an `acceptsConfig(Map<String, dynamic> configMap)` method to the `IntegrationParty` class, that returns `true`, if this integration party can use this config, and `false` otherwise.
4. Add a `getParty(Map<String, dynamic> configMap)` method to the `Parties` abstract class. This method returns an `IntegrationParty` that can use this config or throws an `UnimplementedError` if the party that accepts the given config is not found.
5. Create a `ConfiguredParty` that holds the `IntegrationParty` and the `Config` this party accepts. Create `ConfiguredSourceParty` and `ConfiguredDestinationParty` that extend the `ConfiguredParty` abstract class.
6. Create a `ConfiguredParties` class that holds source and destination configured parties.
7. Create a `ConfiguredPartiesFactory` class that creates the `ConfiguredParties` from the `RawIntegrationConfig` and the `SupportedIntegrationParties`.

> ## Comparison Of Old And New Approaches

Reviewer: ```suggestion
## Approaches Comparison
```
What do you think?

> Consider the following code snippets that show how the proposed changes may simplify the codebase, assuming that the desired code goal is to parse the source and destination `IntegrationParties` with their specific `Config`s.

Reviewer: ```suggestion
Consider the following code snippets that show how the proposed changes may simplify the codebase, assuming that the desired code goal is to get the source and destination `IntegrationParties` and parse their specific `Config`s.
```

  }

  T parseConfig<T extends Config>(
    Map<String, dynamic> configMap,
    IntegrationParty<T, IntegrationClient> party,
  ) {
    final config = party.configParser.parse(configMap);
    logger.info('$config was created.');

    return config;
  }
```

### After

``` dart
  ValidateCommand({
    this.rawIntegrationConfigFactory,
    this.configuredPartiesFactory,
  });

  @override
  Future<void> run() async {
    final configFilePath = getArgumentValue('config-file-path') as String;

    try {
      final rawIntegrationConfig = rawIntegrationConfigFactory.create(configFilePath);

      final configuredParties = configuredPartiesFactory.create(rawIntegrationConfig);

      final configuredSourceParty = configuredParties.configuredSourceParty;
      final configuredDestinationParty = configuredParties.configuredDestinationParty;

      final sourceParty = configuredSourceParty.party;
      final sourceConfig = configuredSourceParty.config;
      // specific logic here

      final destinationParty = configuredDestinationParty.party;
      final destinationConfig = configuredDestinationParty.config;
      // specific logic here
    }
  }
```

## General Class Diagram

> Consider the following class diagram, that describes the required changes to share the code between the `Sync` and the `Validate`commands.

Reviewer: ```suggestion
Consider the following class diagram that describes the required classes and interfaces used to share the code between commands:
```
